File: sources/java/14_file_read/Main.java
// task 14 file_read — expected output: 484442112
// build: javac Main.java    run: java Main
// build (graalvm native-image): native-image -O2 Main    run: ./Main
// Reads data.bin from the working directory in 1 MiB chunks; bytes are unsigned via & 0xFF.

import java.io.FileInputStream;
import java.io.IOException;

public class Main {
    public static void main(String[] args) throws IOException {
        byte[] buf = new byte[1024 * 1024];
        long total = 0;
        try (FileInputStream in = new FileInputStream("data.bin")) {
            int n;
            while ((n = in.read(buf)) != -1) {
                for (int i = 0; i < n; i++) {
                    total += buf[i] & 0xFF;
                }
            }
        }
        System.out.println(total % 4294967296L);
    }
}

File: sources/cpp/14_file_read/main.cpp
// task 14 file_read — expected output: 484442112
// build: g++ -O2 -pthread -o prog main.cpp    run: ./prog
// also builds with: clang++ -O2 -pthread -o prog main.cpp | cl /O2 /EHsc /Fe:prog main.cpp
// note: reads data.bin from the working directory in 1 MiB chunks and prints the byte sum
//       reduced modulo 4294967296.

#include <cstdio>

int main() {
    static unsigned char buffer[1024 * 1024];

    std::FILE* file = std::fopen("data.bin", "rb");
    if (file == nullptr) {
        return 1;
    }

    unsigned long long total = 0;
    size_t got = 0;
    while ((got = std::fread(buffer, 1, sizeof(buffer), file)) > 0) {
        for (size_t i = 0; i < got; ++i) {
            total += static_cast<unsigned long long>(buffer[i]);
        }
    }

    std::fclose(file);

    std::printf("%llu\n", total % 4294967296ULL);
    return 0;
}
